php/Test/indexTest.php

```diff
@@ -0,0 +1,145 @@
+    $this->assertEquals($json_hash_response, $response['response']);
+  }
+
+  public function testErrorHandlingFromClever() {
+    $mock_request_options = prepare_options_for_clever();
+    $this->http->mock->
+    when()->methodIs('GET')->pathIs('/')->
+    then()->statusCode(500)->end();
+    $this->http->setUp();
+    $response = request_from_clever("http://localhost:1234", array(), $mock_request_options);
+    $this->assertEquals(500, $response['response_code']);
+  }
+
+  public function testErrorHandlingFromCurl() {
+    $mock_request_options = prepare_options_for_clever();
+    try {
+      $response = request_from_clever("httpl://explorer:1234", array(), $mock_request_options);
+      $this->fail("Expected exception not thrown");
+    } catch(Exception $e) {
+      $this->assertRegExp("@cURL failure@", $e->getMessage());
+    }
+  }
+
+
+  public function testGenerateSignInWithCleverUrl() {
+    $our_options = prepare_options_for_clever();
+    $wanted_redirect_url = preg_quote(urlencode($our_options['client_redirect_url']));
+    $wanted_client_id = preg_quote($our_options['client_id']);
+    $wanted_district_id = preg_quote($our_options['district_id']);
+    $wanted_authorize_url = $our_options['clever_oauth_authorize_url'];
+    $url = generate_sign_in_with_clever_url($our_options);
+    $this->assertStringStartsWith($wanted_authorize_url, $url);
+    $this->assertRegExp("@redirect_uri=$wanted_redirect_url@", $url);
+    $this->assertRegExp(("@district_id=$wanted_district_id@"), $url);
+    $this->assertRegExp(("@client_id=$wanted_client_id@"), $url);
+  }
+
+  public function testGenerateSignInWithCleverLink() {
+    $our_options = prepare_options_for_clever();
+    $url = generate_sign_in_with_clever_url($our_options);
+    $link = generate_sign_in_with_clever_link($our_options);
+    $this->assertStringStartsWith("<a href='$url'", $link);
+    $this->assertStringEndsWith("</a>", $link);
+  }
+
+  public function testProcessIncomingRequests() {
+    $our_options = prepare_options_for_clever();
+  }
+}
+
+# Instructs our API requests to use mock environment
+function prepare_options_for_clever() {
+  $options = array(
+    "client_redirect_url" => "http://localhost:1234/oauth",
+    "client_id" => 'abc',
+    "client_secret" => 'def',
+    "district_id" => '123',
+    "port" => 1234,
+    'clever_oauth_tokens_url' => "http://localhost:1234/oauth/tokens",
+    'clever_api_me_url' => "http://localhost:1234/me",
+    "clever_oauth_authorize_url" => "http://localhost:1234/oauth/authorize",
+  );
+  return $options;
+}
+
+# Prepares a consistent response to /me
+function prepare_me_response_hash() {
+  return array(
+    'data' => array(
+      "district" => "4fd43cc56d11340000000005",
+      "id" => "4fee004cca2e43cf2700028b",
+      "type" => "student"
+    ),
+    'links' => array(
+      array(
+        'uri' => "/me",
+        'rel' => "self"
+      )
+    )
+  );
+}
+?>
```

Comment:
should ditch these file endings, they're not required unless you're inlineing into html and can lead to unexpected WS injection  .